k8s: Make agent wait for CRDs

christarazi opened the pull request:
See commit msgs.

```release-note
Cilium Agent will now wait for CRDs to become available instead of the Operator; the Operator will register the CRDs
```

christarazi committed:
daemon, k8s: Make agent wait for CRDs

Previously, cilium-operator was responsible for waiting for the CRDs to
be registered because the agent was registering them. Now the roles have
been inverted.

This commit implements a new CRD controller that will fetch the CRDs
efficiently, by requesting them as metav1.PartialObjectMetadata. These
objects contain the minimal metadata to identify the object, without
containing the full validation schema, for example, in the case of CRDs.
This is important because moving the waiting to the agent, from
cilium-operator, means that there are many more agents that will hit the
K8s apiserver upon startup, than cilium-operators would, due
cilium-agent being a Daemonset. The efficient retrieval of the CRDs will
help minimize the load on the apiserver when the agents start up.

This commit effectively reverts 8b4b01009 ("operator: Revert waiting for
CRDs"), but introduces the functionality in a complete different manner.

In addition, this commit reorders the agent (daemon) initialization steps to
first wait until all Cilium CRDs have been registered. It is important to
perform this first before the CiliumNode and other node initialization code
runs, as specific node information is needed when the Services and Pods
controllers begin. Once the node initialization code is run, then we can kick
off all the K8s watchers (controllers). This also prevents the agent from
blocking its restart while it's waiting for K8s subsys to be initialized.

Co-authored-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon: Move CRD wait timeout option to agent

Previously, this option existed in cilium-operator. It has been
deprecated in 325547f6e ("operator: Deprecate crd-wait-timeout").

This commit moves this option to the agent. In a future commit, the
agent will be taking on the resposbility of waiting for the CRDs to be
registered. Previously, cilium-operator had this resposbility.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Refactor K8s client code

This commit only contains non-functional changes. It is a refactor meant
to ease the readability of the file. The highlight of the changes:

 - Rearrange functions from most important to least important
 - Unexport CreateClient() because it was not used outside of this file
 - Rename vars and functions to be consistent with Golang style
 - Fix Godocs on each exported type

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Define and use dedicated apiextensions client

This helps code reuse. This client is used for registering the CRDs and
to wait for the CRDs.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Fix variable name of CLRP CRD

This is a cosmetic change.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -89,6 +89,10 @@ func (k *K8sWatcher) ciliumEndpointsInit(ciliumNPClient *k8s.K8sCiliumClient, as
 		once.Do(func() {
+			if err := k8s.WaitForCRD(k8s.APIExtClient(), cilium_v2.CEPName); err != nil {
+				log.WithError(err).Fatalf("Unable to find %s CRD", cilium_v2.CEPName)
+			}
```
These wait operations are blocking, running before each watcher is run. Should we consider trying to move the waiting mechanism so that it's not blocking and parallelize it?

aanm replied:
The `WaitForCRD` creates a watcher for each call, which we want to avoid. Instead, we can reuse the `k.blockWaitGroupToSyncResources` and in here we can have `k.WaitForCacheSync("CRD"+cilium_v2.CEPName)`

aanm commented on pkg/k8s/init.go:
```diff
@@ -271,3 +280,101 @@ func RegisterCRDs() error {
+				return client.ApiextensionsV1().CustomResourceDefinitions().List(ctx, options)
+			}
+
+			return client.ApiextensionsV1beta1().CustomResourceDefinitions().List(ctx, options)
```
we should make the `client.ApiextensionsV1beta1` conditional based on the k8s version that supports the v1 vs v1beta1.

christarazi replied:
Not sure if I'm following, but we are doing this a few lines above.

aanm replied:
:man_facepalming: for some reason I read `ApiextensionsV1beta1` on both of them.

aanm commented on pkg/k8s/init.go:
```diff
@@ -271,3 +280,101 @@ func RegisterCRDs() error {
+func waitForCRD(ctx context.Context, client clientset.Interface, name string) error {
+	log.Infof("Waiting for CRD %s to be available", name)
+
+	selector := fields.OneTermEqualSelector("metadata.name", name).String()
```
instead of creating a single "waitForCRD" for each CRD we should watch for all CRDs. We should also make sure that the watcher does not download the entire CRD for all agents.

In curl it's easy to do it by passing this headers (specially the `as=Table`):
```
curl -k -v -XGET  -H "Accept: application/json;as=Table;v=v1;g=meta.k8s.io,application/json;as=Table;v=v1beta1;g=meta.k8s.io,application/json" -H "User-Agent: kubectl/v1.19.1 (linux/amd64) kubernetes/206bcad" 'https://192.168.33.11:6443/apis/apiextensions.k8s.io/v1/customresourcedefinitions?resourceVersion=1034518&watch=true'
```

otherwise all cilium agents will download the CRD which is unnecessary since we only care about its name

christarazi replied:
Synced offline. The gist was that we needed to implement a custom controller of sorts in order to retrieve CRDs in an efficient manner.

joestringer reviewed (approved):
CLI LGTM

christarazi committed:
k8s: Allow hasSynced func to be passed directly

A hasSynced function is part of a controller which returns a boolean
indicating whether it has synced all the K8s resources it was watching
to its cache.

This is a preparatory commit for a future commit that will create a CRD
controller with a custom hasSynced function, rather than depending on
the underlying K8s controller implementation. The reason is because we
needed to customize how the CRD controller retrieves CRDs for efficiency
purposes, which are important in large-scale environments.

Co-authored-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,294 @@
+
+	// The above loop will call blockWaitGroupToSyncResources to populate the
+	// K8sWatcher state with the current state of the CRDs. It will check the
+	// state of each CRD, with the function provided below. If the function
```
```suggestion
	// state of each CRD, with the inline function provided. If the function
```

christarazi commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,294 @@
+	// in the cluster). If (1) is encountered, then Cilium will fatal because
+	// it cannot proceed if the CRDs are not present. If (2) is encountered,
+	// then make sure the controller has exited by cancelling the context and
+	// we return out.
```
```suggestion
	// Meanwhile, the Goroutine below kicks off the controller that was
	// instantiated above, and enters a loop looking for (1) if the context has
	// deadlined or (2) if the entire CRD state has been synced (all CRDs found
	// in the cluster). While the Goroutine is in this loop, the controller is listening
	// for either add or delete events to the customresourcedefinition resource 
	// (disguised inside a metav1.PartialObjectMetadata object).
	// If (1) is encountered, then Cilium will fatal because
	// it cannot proceed if the CRDs are not present. If (2) is encountered,
	// then make sure the controller has exited by cancelling the context and
	// we return out.
```

christarazi commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,294 @@
+					log.Info("All Cilium CRDs have been found and are available")
+					return
+				}
+			}
```
Looking for feedback on if we can do this in a more efficient way. I could only think of polling...

joestringer replied:
It's not the worst thing in the world to poll this every second. I guess a cleaner approach would be to have a channel to check whether all the CRDs were synced, which would be tied into `crdState`. Then the logic inside `isSynced()` would run somewhere in a separate goroutine to close the "CRDsSynced" channel when the current logic returns true.

Though I could easily see the ticking just being moved to that goroutine. Then one step further would be to use channels all the way up I guess..

christarazi replied:
Thanks. Yeah it doesn't seem like it's possible to do this without polling in someway. Whether the polling happens directly in the loop or inside a helper goroutine somewhere else, the functionality is the same. I am keeping the current approach as adding another channel just adds extra complexity and indirection that doesn't provide much benefit, besides maybe "elegance". I like the directness of the code currently. :)

joestringer replied:
Agree, I think that's the right tradeoff.

christarazi commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,294 @@
+			// PartialObjectMetadata.
+			fmt.Printf("chris received %T, expected *metav1.PartialObjectMetadata raw %s\n",
+				row.Object.Object, string(row.Object.Raw))
+			continue
```
Maybe logging an info msg? I don't think we want to halt in any way if we encounter an error here.

aanm replied:
Yes, I think we should log an `Error` because if we get something that we aren't expecting Cilium will wait for the CRDs forever.

christarazi commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -448,6 +458,10 @@ func (k *K8sWatcher) EnableK8sWatcher(queueSize uint) error {
 	asyncControllers := &sync.WaitGroup{}
 
+	// Cilium CRDs
+	fmt.Println("chris launching crd k8s watcher")
```
Remove

aanm commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,299 @@
+	s.Lock()
+	defer s.Unlock()
+	for crd := range s.m {
+		if s.m[crd] == false {
```
small nit, you can access the value if you do `for _, ready := range s.m { if !ready {return false}}`

aanm commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,299 @@
+}
+
+const (
+	cnpCRD  = "crd:" + v2.CNPName
```
Create a helper function to join the `crd:` with the name and it can be used in [1]

aanm commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,299 @@
+			AddFunc: func(obj interface{}) {
+				if pom, ok := obj.(*metav1.PartialObjectMetadata); ok {
+					crds.Lock()
+					crds.m["crd:"+pom.GetName()] = true
```
create a helper function in `addCiliumNetworkPolicyV2`

aanm commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,299 @@
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if pom, ok := obj.(*metav1.PartialObjectMetadata); ok {
```
Use a helper function to translate from `obj` to `*metav1.PartialObjectMetadata`. The `obj` might be a `cache.DeletedFinalStateUnknown`. See the other functions in `pkg/k8s/factory_functions.go`

aanm commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,299 @@
+	// Create a context so that we can timeout after the configured CRD wait
+	// peroid.
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.CRDWaitTimeout)
+	for crd := range crds.m {
```
you need to protect `crds.m` while doing this loop :-)

aanm commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,299 @@
+// tableToPomList converts a metav1.Table's cells into individual
+// metav1.PartialObjectMetadata objects, and returns them all placed in a
+// metav1.PartialObjectMetadataList.
+func tableToPomList(t *metav1.Table) *metav1.PartialObjectMetadataList {
```
Small nit. This is a `metav1.Table` -> `metav1.POM`. We likely need another function that does `v1beta1.Table` to `v1beta1.POM`. Unless the v1.Table is available since k8s >= 1.12

christarazi replied:
Addressed by putting a comment on why we may not need to do this.

aanm commented on pkg/k8s/watchers/crd.go:
```diff
@@ -0,0 +1,299 @@
+							"%v timeout. Please ensure that Cilium Operator is "+
+							"running, as it's responsible for registering all "+
+							"the Cilium CRDs.",
+							option.Config.CRDWaitTimeout)
```
might be worth presenting which CRDs are **not** registered.

christarazi committed:
k8s: Add capability flag for watching metav1.POM

The ability to watch metav1.PartialObjectMetadata (or POM) and
metav1.Table was introduced in K8s 1.15. [1]

This is relevant because our CRD controller attempts to fetch CRDs in
the cluster efficiently by requesting the CRD in a POM object. The CRD
controller does this in order to avoid requesting the full object, which
may contain a large validation schema and other irrelevant fields. This
is important because in large-scale environments, all agents will
request all the CRDs at once which will put unnecessary load on the
apiserver.

However, we cannot perform this request at all on versions of K8s
1.14 and below. Therefore, we must fullback to requesting the full CRD
object. Hence, this commit allows us to check whether the apiserver
supports this action, so that we can efficiently request CRDs on
versions that do support it.

[1]: KEP:
https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190322-server-side-get-to-ga.md#goals

[1]: PR: https://github.com/kubernetes/kubernetes/pull/71548

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Add ability to watch for full CRD object

This fixes the CRD controller to work for K8s 1.14 and below.

Fixes: https://github.com/cilium/cilium/issues/13498

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm committed:
k8s: install all CRDs regardless of options used

To avoid complexity in the daemon, the operator should install all CRDs
regardless of the options used in the operator. This prevents Cilium
from stopping its initialization while waiting for the CRD
"ciliumidentities.cilium.io".

Signed-off-by: André Martins <andre@cilium.io>

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1365,9 +1370,11 @@ func runDaemon() {
+		// Wait only for certain caches, but not all!
+		// (Check Daemon.initK8sSubsystem() for more info)
+		<-d.k8sCachesSynced
+	}
```
This change should be part of the previous commit. If the CI passes there is not need to re-trigger the CI after moving this change to the previous commit.

brb reviewed (changes requested):
Re ` daemon/cmd: move k8s subsys initialization before`: the NodePort addr initialization cannot happen before `detectNativeDevices()` is called. Meanwhile, the latter depends on `pkg/k8s/init.go:WaitForNodeInformation()`.

christarazi committed:
daemon: Fix receive on K8s cache channel in init

The channel should only be received from if K8s is enabled, otherwise it
will block forever. Once the K8s resources are synced, the channel will
be closed and unblocks.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1373,9 +1373,11 @@ func runDaemon() {
 	bootstrapStats.enableConntrack.End(true)
 
 	bootstrapStats.k8sInit.Start()
```
`bootstrapStats.k8sInit` could be inside the `k8s.IsEnabled()`

aanm commented on daemon/cmd/daemon.go:
```diff
@@ -461,6 +461,12 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
+		// context deadline or if the context has been cancelled, the context's
+		// error will be returned. Otherwise, it succeeded.
+		if err := d.k8sWatcher.WaitForCRDsToRegister(d.ctx); err != nil {
+			return nil, restoredEndpoints, err
```
small nit, this could be `return nil, nil, err`
